From 0b5b1be810baf881d835ef2a96d55f54f537a6a5 Mon Sep 17 00:00:00 2001
From: Mark Thomas <markt@apache.org>
Date: Wed, 2 Nov 2016 21:24:45 +0000
Subject: [PATCH] Fix DBCP-472

Avoid potential infinite loops when checking if an SQLException is fatal
for a connection or not.
---
 src/changes/changes.xml                                       | 4 ++++
 .../java/org/apache/commons/dbcp2/PoolableConnection.java     | 3 ++-
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 72d368d20e..c03667c96f 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -95,6 +95,10 @@ The <action> type attribute can be add,update,fix,remove.
         Correct the name of the configuration attribute
         softMinEvictableIdleTimeMillis.
       </action>
+      <action dev="markt" type="fix" issue="DBCP-472">
+        Avoid potential infinite loops when checking if an SQLException is fatal
+        for a connection or not.
+      </action>
     </release>
     <release version="2.1.1" date="6 Aug 2015" description=
 "This is a patch release, including bug fixes only.">
diff --git a/src/main/java/org/apache/commons/dbcp2/PoolableConnection.java b/src/main/java/org/apache/commons/dbcp2/PoolableConnection.java
index 2773f4f691..7424240f69 100644
--- a/src/main/java/org/apache/commons/dbcp2/PoolableConnection.java
+++ b/src/main/java/org/apache/commons/dbcp2/PoolableConnection.java
@@ -326,7 +326,8 @@ private boolean isDisconnectionSqlException(final SQLException e) {
             fatalException = _disconnectionSqlCodes == null ? sqlState.startsWith(Utils.DISCONNECTION_SQL_CODE_PREFIX)
                     || Utils.DISCONNECTION_SQL_CODES.contains(sqlState) : _disconnectionSqlCodes.contains(sqlState);
             if (!fatalException) {
-                if (e.getNextException() != null) {
+                SQLException nextException = e.getNextException();
+                if (nextException != null && nextException != e) {
                     fatalException = isDisconnectionSqlException(e.getNextException());
                 }
             }
